File: java/src/main/java/com/penapereira/patterns/producerconsumer/Producer.java
package com.penapereira.patterns.producerconsumer;

import java.util.List;

/** Puts its items into the buffer in order; put() waits while the buffer is full. */
public final class Producer implements Runnable {

    private final BoundedBuffer<Integer> buffer;
    private final List<Integer> items;

    public Producer(BoundedBuffer<Integer> buffer, List<Integer> items) {
        this.buffer = buffer;
        this.items = items;
    }

    @Override
    public void run() {
        try {
            for (int item : items) {
                buffer.put(item);
            }
        } catch (InterruptedException stop) {
            Thread.currentThread().interrupt();
        }
    }
}

File: typescript/src/producer-consumer/producer.ts
import type { BoundedBuffer } from "./bounded-buffer.ts";
import type { Task } from "./task.ts";

/** Puts its items into the buffer in order. */
export class Producer {
  constructor(
    private readonly buffer: BoundedBuffer<number>,
    private readonly items: readonly number[],
  ) {}

  *run(): Task {
    for (const item of this.items) {
      yield* this.buffer.put(item);
    }
  }
}
